### PR TITLE
Improve support for -Xobjc-generics and enable it by default

### PR DESCRIPTION
**Pull request by SvyatoslavScherbina**

### CODE REVIEW

File: OBJC_INTEROP.md [hunk 1]
```diff
@@ -229,26 +229,26 @@ foo {
 Objective-C supports "lightweight generics" defined on classes, with a relatively limited feature set. Swift can import 
 generics defined on classes to help provide additional type information to the compiler.
 
-Generic feature support for Objc and Swift differ from Kotlin, so the translation will inevitably lose some information,
+Generic feature support for Obj-C and Swift differ from Kotlin, so the translation will inevitably lose some information,
```
Comment (olonho) [inline OBJC_INTEROP.md]: Objective-C

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/objcexport/ObjCExportNamer.kt [hunk 2]
```diff
@@ -161,6 +168,12 @@ private class ObjCExportNamingHelper(
         "Type" -> "${name}_" // See https://github.com/JetBrains/kotlin-native/issues/3167
         else -> name
     }
+
+    fun isTypeParameterNameReserved(name: String): Boolean = name in reservedTypeParameterNames
+
+    private val reservedTypeParameterNames = setOf("id", "NSObject", "NSArray", "NSCopying", "NSNumber", "NSInteger",
```
Comment (olonho) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/objcexport/ObjCExportNamer.kt]: Where this list comes from?

File: backend.native/tests/framework/stdlib/stdlib.swift [hunk 3]
```diff
@@ -216,10 +216,10 @@ class StdlibTests : TestProvider {
         try applyVoid { $0.add("bar") }
         try applyVoid { $0.remove("baz") }
         try applyVoid { $0.add("baz") }
-        try applyVoid { $0.add(TripleVals(first: 1, second: 2, third: 3)) }
-        try apply { $0.member(TripleVals(first: 1, second: 2, third: 3)) as! TripleVals }
+        try applyVoid { $0.add(TripleVals<NSNumber>(first: 1, second: 2, third: 3)) }
```
Comment (olonho) [inline backend.native/tests/framework/stdlib/stdlib.swift]: Just curious, why it is needed to specify `NSNumber`?

File: backend.native/tests/framework/stdlib/stdlib.swift [hunk 4]
```diff
@@ -314,9 +314,9 @@ class StdlibTests : TestProvider {
     }
 
     func testSet() throws {
-        let elements = KotlinArray(size: 2) { index in nil }
+        let elements = KotlinArray<AnyObject>(size: 2) { index in nil }
         elements.set(index: 0, value: nil)
-        elements.set(index: 1, value: 42)
+        elements.set(index: 1, value: 42 as NSNumber)
```
Comment (olonho) [inline backend.native/tests/framework/stdlib/stdlib.swift]: Is it always required now?

File: OBJC_INTEROP.md [hunk 5]
```diff
@@ -327,6 +312,20 @@ In Kotlin you can provide upper bounds for a generic type. Objective-C also supp
 in more complex cases, and is currently not supported in the Kotlin - Objective-C interop. The exception here being a non-null
 upper bound will make Objective-C methods/properties non-null.
 
+### To disable
+
+To have the framework header written without generics, add a flag to the compiler config:
```
Comment (olonho) [inline OBJC_INTEROP.md]: `add the flag`? It's precisely *that* flag.

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/objcexport/ObjCExportNamer.kt (hunk 2 above)
Comment (SvyatoslavScherbina) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/objcexport/ObjCExportNamer.kt]: Generally from here: https://github.com/JetBrains/kotlin-native/pull/3778/files#diff-d13ae8cff625fe7530e3a86b4102ca0fL666

I have some plans to improve this part of code, but it requires more deep changes.

File: backend.native/tests/framework/stdlib/stdlib.swift (hunk 3 above)
Comment (SvyatoslavScherbina) [reply backend.native/tests/framework/stdlib/stdlib.swift]: Swift compiler infers `Int` as type argument, but Objective-C classes can have only "class" (e.g. reference) types as type arguments.

File: backend.native/tests/framework/stdlib/stdlib.swift (hunk 4 above)
Comment (SvyatoslavScherbina) [reply backend.native/tests/framework/stdlib/stdlib.swift]: Not exactly. Swift doesn't actually have implicit conversion from its integer types to `NSNumber`. There is ad hoc support for literals, e.g.
```swift
let x: NSNumber = 42
```
works.
But in the discussed line of test the expected type is not `NSNumber`, so such a literal conversion doesn't apply.

File: OBJC_INTEROP.md (hunk 5 above)
Comment (SvyatoslavScherbina) [reply OBJC_INTEROP.md]: Fixed. Btw, this sense was originally written by a native speaker `¯\_(ツ)_/¯`

File: OBJC_INTEROP.md (hunk 1 above)
Comment (SvyatoslavScherbina) [reply OBJC_INTEROP.md]: Fixed.

**Pull request merged**